File: leetcode/num295.py
class MedianFinder(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.nums = []
        self.length = 0

    def addNum(self, num):
        """
        :type num: int
        :rtype: None
        """
        if self.length == 0:
            self.nums.append(num)
            self.length += 1
        else:
            index = self.length-1
            for i in range(self.length):
                if num <= self.nums[i]:
                    index = i
                    break
            self.nums.insert(index, num)
            self.length += 1

    def findMedian(self):
        """
        :rtype: float
        """
        if self.length % 2 != 0:
            return self.nums[int(self.length/2)]
        else:
            return (self.nums[int(self.length/2)] + self.nums[int(self.length/2)-1]) * 0.5

# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
if __name__ == '__main__':
    mf = MedianFinder()
    mf.addNum(-1)
    print(mf.findMedian())
    mf.addNum(-2)
    print(mf.findMedian())
    mf.addNum(-3)
    print(mf.findMedian())
    mf.addNum(-4)
    print(mf.findMedian())
    mf.addNum(-5)
    print(mf.nums)
    print(mf.findMedian())

File: leetcode/num295.java
package leetcode;
import java.util.PriorityQueue;
import java.util.Queue;

class MedianFinder {
    private Queue<Integer> small;
    private Queue<Integer> large; // 从小到大，存放右半边偏大的数
    int count = 0;

    /** initialize your data structure here. */
    public MedianFinder() {
        small = new PriorityQueue<Integer>((i1, i2) -> (i2 - i1));
        large = new PriorityQueue<Integer>(); // 从小到大，存放右半边偏大的数
    }

    public void addNum2(int num) {
        if (count % 2 == 0)
            small.offer(num);
        else
            large.offer(num);
        if (!small.isEmpty() && !large.isEmpty() && small.peek() > large.peek()) {
            small.offer(large.poll());
            large.offer(small.poll());
        }
        count += 1;
    }

    /**
     * 始终维持small.size > large.size
     * @param num
     */
    public void addNum(int num) {
        small.offer(num);
        large.offer(small.poll());
        if (large.size() > small.size()){
            small.offer(large.poll());
        }
    }

    public double findMedian() {
        return small.size() > large.size() ? small.peek() : (small.peek() + large.peek()) / 2.0;
    }
}